OCPBUGS-55211: use regular user to create httproute

lihongan opened the pull request:
enhancement for httproute test

1. use regular user to create httproute as well as pod/service
2. remove the step of cleaning up httproute

lihongan commented:
/jira refresh

lihongan commented on test/extended/router/gatewayapicontroller.go:
```diff
@@ -324,12 +315,12 @@ func createAndCheckGateway(oc *exutil.CLI, gwname, gwclassname, domain string) (
 	waitErr := wait.PollUntilContextTimeout(context.Background(), 2*time.Second, 10*time.Minute, false, func(context context.Context) (bool, error) {
-		gateway, errGwStatus := gwapiClient.GatewayV1().Gateways(ingressNameSpace).Get(context, gwname, metav1.GetOptions{})
-		if errGwStatus != nil {
+		gateway, err = oc.AdminGatewayApiClient().GatewayV1().Gateways(ingressNameSpace).Get(context, gwname, metav1.GetOptions{})
```
Should not use `:=` here, otherwise the `gateway.Name` is empty in logs as below
```
I0422 15:27:59.859132 34018 gatewayapicontroller.go:352] Gateway "" successfully programmed!
```

lihongan replied:
@melvinjoseph86 please check your enhancement PR, not sure why you revert the fix.

lihongan commented on test/extended/router/gatewayapicontroller.go:
```diff
@@ -347,15 +338,15 @@ func checkGatewayStatus(oc *exutil.CLI, gwname, ingressNameSpace string) (*gatew
 // buildGateway initializes the Gateway and returns its address.
 func buildGateway(name, namespace, gcname, fromNs, domain string) *gatewayapiv1.Gateway {
-	hostname := gatewayapiv1.Hostname("*." + "gwapi." + domain)
+	hostname := gatewayapiv1.Hostname("*." + domain)
```
Needn't to insert `gwapi.` to the hostname, otherwise it looks like
```
  spec:
    gatewayClassName: openshift-default
    listeners:
    - allowedRoutes:
        namespaces:
          from: All
      hostname: '*.gwapi.gw-default.ci-ln-3if6zbk-72292.origin-ci-int-gce.dev.rhcloud.com'
```
Indeed, we replace `apps.` to `gw-xxx.` in the test to generate the domain for gateway.

lihongan commented:
/test e2e-gcp-ovn-techpreview

lihongan commented:
/test e2e-gcp-ovn-techpreview
/retest-required

lihongan commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-techpreview periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-dualstack-techpreview periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview periodic-ci-openshift-multiarch-master-nightly-4.19-ocp-e2e-aws-ovn-techpreview-multi-a-a periodic-ci-openshift-release-master-nightly-4.19-e2e-vsphere-ipi-ovn-runc-techpreview

alebedev87 commented on test/extended/router/gatewayapicontroller.go:
```diff
@@ -231,31 +206,50 @@ var _ = g.Describe("[sig-network-edge][OCPFeatureGate:GatewayAPIController][Feat
+	o.Expect(err).NotTo(o.HaveOccurred())
+	o.Expect(infra).NotTo(o.BeNil())
+
+	platformType := infra.Status.Platform
+	if infra.Status.PlatformStatus != nil {
+		platformType = infra.Status.PlatformStatus.Type
+	}
```
Just noticed that we set the platform type with the deprecated field:
```
    // platform is the underlying infrastructure provider for the cluster.
    //   
    // Deprecated: Use platformStatus.type instead.
    Platform PlatformType `json:"platform,omitempty"`

    // platformStatus holds status information specific to the underlying
    // infrastructure provider.
    // +optional
    PlatformStatus *PlatformStatus `json:"platformStatus,omitempty"`
```
I suppose that if `infra.Status.PlatformStatus==nil`  then `infra.Status.Platform` can be of no use too (e.g. empty).

We can set `platformType := ""` and add `o.Expect()` to check that we got the platform type from `infra.Status.PlatformStatus.Type`.

lihongan replied:
Change as below
```
platformType := infra.Status.PlatformStatus.Type
o.Expect(platformType).NotTo(o.BeNil())
```

lihongan commented on test/extended/router/gatewayapicontroller.go:
```diff
@@ -231,31 +206,50 @@ var _ = g.Describe("[sig-network-edge][OCPFeatureGate:GatewayAPIController][Feat
 		g.By("Create the http route using the custom gateway")
-		httproutes = append(httproutes, "test-httproute")
-		defaultRoutename := "test-hostname.gwapi." + customDomain
+		defaultRoutename := "test-hostname." + customDomain
```
Removing the `gwapi.` to match the change in `func buildGateway()`

alebedev87 reviewed (commented):
LGTM overall, just a couple of nit picks. Do you plan to hold the merge of this PR until https://github.com/openshift/api/pull/2284 is merged?

lihongan commented on test/extended/util/client.go:
```diff
@@ -674,6 +675,10 @@ func (c *CLI) BuildClient() buildv1client.Interface {
 	return buildv1client.NewForConfigOrDie(c.UserConfig())
 }
 
+func (c *CLI) GatewayApiClient() gatewayapiv1client.Interface {
+	return gatewayapiv1client.NewForConfigOrDie(c.UserConfig())
+}
```
Yes, add comment `// GatewayApiClient provides a GatewayAPI client for the current namespace user.` for this,

lihongan commented on test/extended/util/client.go:
```diff
@@ -717,6 +722,10 @@ func (c *CLI) AdminConfigClient() configv1client.Interface {
 	)
 }
 
+func (c *CLI) AdminGatewayApiClient() gatewayapiv1client.Interface {
+	return gatewayapiv1client.NewForConfigOrDie(c.AdminConfig())
+}
```
Added comment `// AdminGatewayApiClient provides a GatewayAPI client for the cluster admin user.` for this one, the user is cluster admin and has all rights for Gateway API CRDs.
The other client is for current namespace user, it looks like `e2e-test-gatewayapi-controller-rbcwp-user` in CI.

lihongan commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-techpreview periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-dualstack-techpreview periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview periodic-ci-openshift-release-master-nightly-4.19-e2e-vsphere-ipi-ovn-runc-techpreview periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview periodic-ci-openshift-release-master-ci-4.19-e2e-azure-ovn-techpreview

alebedev87 commented on test/extended/router/gatewayapicontroller.go:
```diff
@@ -231,31 +206,48 @@ var _ = g.Describe("[sig-network-edge][OCPFeatureGate:GatewayAPIController][Feat
+	o.Expect(err).NotTo(o.HaveOccurred())
+	o.Expect(infra).NotTo(o.BeNil())
+
+	platformType := infra.Status.PlatformStatus.Type
+	o.Expect(platformType).NotTo(o.BeNil())
```
`infra.Status.PlatformStatus` can be `nil` which may cause a `panic` during dereferencing.
```suggestion
	o.Expect(infra.Status.PlatformStatus).NotTo(o.BeNil())
	platformType := infra.Status.PlatformStatus.Type
	o.Expect(platformType).NotTo(o.BeEmpty())
```

lihongan replied:
Good catch! Thanks for the suggestion.

lihongan commented:
/test e2e-gcp-ovn-techpreview

lihongan commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-techpreview periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-dualstack-techpreview periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview periodic-ci-openshift-release-master-nightly-4.19-e2e-vsphere-ipi-ovn-runc-techpreview periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview periodic-ci-openshift-release-master-ci-4.19-e2e-azure-ovn-techpreview

lihongan commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview periodic-ci-openshift-release-master-nightly-4.19-e2e-vsphere-ipi-ovn-runc-techpreview periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview periodic-ci-openshift-release-master-ci-4.19-e2e-azure-ovn-techpreview

lihongan commented:
Some clusters (e.g. metal ipv6) are disconnected and catalogsources are disabled as below, so OSSM cannot be installed. Consider the complexity I prefer the old way and skip all tests if non-cloud platform.  
```
$ oc get operatorhub cluster -oyaml
<......>
spec:
  disableAllDefaultSources: true
status:
  sources:
  - disabled: true
    name: redhat-operators
    status: Success
<......>
```

alebedev87 commented:
/lgtm

lihongan commented:
/assign @xueqzhan 
Could you please help take a look? Thanks

alebedev87 commented:
/unhold

The branch cut is done this morning (Europe time).

lihongan commented:
Resolved the conflicts and force pushed again, retesting...

lihongan commented:
/jira refresh

lihongan commented:
/jira refresh

lihongan commented:
/assign @Miciah 
Could you please help take another look since Andrey takes PTO. Thanks

lihongan commented:
/retest-required

lihongan commented:
/retest-required

lihongan commented:
/retest-required

alebedev87 commented:
/lgtm

lihongan commented:
/retest-required

two required jobs failed due to timed out
```
Process did not finish before 4h0m0s timeout
```

xueqzhan commented:
Is this a concern: https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-origin-29697-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview/1915402680008708096?

lihongan commented:
> Is this a concern: https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-origin-29697-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview/1915402680008708096?

Aah, In the initial commit I'd like to skip 2 tests on Metal and enable other 3 tests on Metal, but I also found the issues you liked here, so finally I reverted the change and keep the original logic (skip all 5 tests on Metal). 
Anyway, I could re-run the payload-job and check the results again.
I also added a [comment](https://issues.redhat.com/browse/OCPBUGS-55211?focusedId=27061665&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-27061665) in the bug to explain this.

lihongan commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview

xueqzhan commented:
/approve
